### PR TITLE
[#9536] Update docs for E2E testing

### PR DESCRIPTION
**Pull request by jtankw3**
Part of #9536

**Outline of Solution**

- Update `development.md` to reflect the changes to testing with the production server as it is not possible to test with the whole test suite anymore because of the login issues
- Add `e2e-testing.md` which describes the general ideas behind e2e testing in TEAMMATES

### CODE REVIEW

File: docs/e2e-testing.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  
+E2E tests in TEAMMATES can be found in the package `teammates.e2e`. 
+
+To set up and run E2E tests, refer to this [document](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md#testing).  
+
+## Creating E2E tests
+  
+As E2E tests should be written from the end user perspective, each test case should reflect some user workflow.   
+  
+In TEAMMATES, E2E test cases are organized by page. Thus, for each page:
```
Comment (wkurniawan07) [inline docs/e2e-testing.md:17]: ~Thus~ For each page, **we**:

File: docs/e2e-testing.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+## Creating E2E tests
+  
+As E2E tests should be written from the end user perspective, each test case should reflect some user workflow.   
+  
+In TEAMMATES, E2E test cases are organized by page. Thus, for each page:  
+1. Identify the important user workflows  
+1. Simulate the user actions involved in the workflow by interacting with the UI elements.  
+1. Assert the expected conditions are present after the interaction.  
+  
+[Selenium](https://www.selenium.dev/) is used to locate and interact with elements in the UI.
```
Comment (wkurniawan07) [inline docs/e2e-testing.md:22]: Another important component for this E2E test is the `BackDoor` class. This should be mentioned somewhere.

File: docs/development.md [hunk 3]
```diff
@@ -246,7 +253,7 @@ If you are testing against a production server (staging server or live server),
 1. Edit `src/e2e/resources/test.properties` as instructed is in its comments.
    * In particular, you will need legitimate Google accounts to be used for testing.
 
-1. Run the full test suite or any subset of it as how you would have done it in dev server.
+1. Login manually to TEAMMATES on the browser used for testing before running any E2E test one at a time. This is required as Google does not allow login by automated software.
```
Comment (wkurniawan07) [inline docs/development.md:256]: I still don't think "one at a time" is a necessary condition. If the reason is because of the tests that are injected with real accounts, then the solution is to make those tests sequential. Reference: <https://github.com/TEAMMATES/teammates/blob/9c60e07c260c3f4fd1b268d627f0cd4b9cda24da/src/test/testng-ci.xml>, take a look at the `sequential-ui-tests` part.

File: docs/e2e-testing.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+- The public methods for page objects should avoid exposing the UI elements it interacts with and instead focus on the functionality of the webpage.   
+- For example, instead of having methods like `fillSearchBox` and `clickSearchButton`, it is better to have a method `searchForInstructor` which hides the UI elements used.    
+
+All Page Object classes inherit from `AppPage` which contains methods that are common for interacting with the web elements such as filling in textboxes.   
+  
+  
+## Things to avoid when writing E2E 
+  
+1. **Testing based on implementation** - The focus should be on user actions instead of implementation details. Therefore, black box testing should be adopted and test cases should be designed around use cases.   
+1. **Exception testing** - Testing edge cases with E2E tests should be avoided. This is because E2E tests are expensive to run and not that effective for isolating bugs. Hence it is not the optimal to use E2E tests to look for exceptions.
```
Comment (wkurniawan07) [inline docs/e2e-testing.md:50]: not ~the~ optimal

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Good starting point, but some more information can be included.
Also, the new document should be mentioned in the `README.md` of the docs directory. Probably under "How-to documents" would be good.

**Comment by wkurniawan07** [comment]
@damithc you may want to review this as well. This is another major paradigm shift from pre-V7 era where the UI tests wrongly (maybe not so back then, but is so right now) attempted to do lower-level tests.

File: docs/development.md (hunk 3 above)
Comment (jtankw3) [reply docs/development.md:256]: It is actually because of this issue where we cannot open two instance of ChromeDriver with the same profile:
https://stackoverflow.com/questions/34727328/run-multiple-chrome-profiles-instances-for-automation-task

And we cannot login without using the profile data (or I have not found a viable solution).

I tried running all e2e tests sequentially but because we do not close the windows of test that fail, the profile becomes locked after one test fails and all subsequent test also fail. I think this can be changed though if we decide that running all the test sequentially for production server e2e is ok.

Firefox is ok.

**Comment by damithc** [comment]
> @damithc you may want to review this as well. This is another major paradigm shift from pre-V7 era where the UI tests wrongly (maybe not so back then, but is so right now) attempted to do lower-level tests.

Nice. @jtankw3 please request a review from me when the contents have settled a bit.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: docs/e2e-testing.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+All E2E test classes inherit from `BaseE2ETestCase` which contains methods that are common to most test cases, such as preparing the `Browser` object used for testing.   
+  
+To help verify the state of the datastore, 
+- `BaseTestCaseWithDatastoreAccess` contains methods to compare datastore entities like `verifyPresentInDatastore`
+- `BaseTestCaseWithBackDoorApiAccess` contains methods that make use of `Backdoor` to access the datastore  
+- `Backdoor` contains methods to create API calls to the back-end without going through the UI. 
+  
+## Page Object Pattern
+  
+In order to make E2E testing more robust to UI changes, the Page Object Pattern is adopted.
```
Comment (damithc) [inline docs/e2e-testing.md:33]: Link to an explanation of the page object pattern?

**Review by damithc — APPROVED** [review]
Looks good 👍 
Good work @jtankw3

File: docs/development.md (hunk 3 above)
Comment (jtankw3) [reply docs/development.md]: For now I have written one at a time for chrome, we can change this at a later date after some testing with the sequential runs? 

Might be good to add a new command for running sequentially or some flag if it works as well.

File: docs/development.md [hunk 6]
```diff
@@ -246,8 +245,12 @@ If you are testing against a production server (staging server or live server),
 1. Edit `src/e2e/resources/test.properties` as instructed is in its comments.
    * In particular, you will need legitimate Google accounts to be used for testing.
 
-1. Run the full test suite or any subset of it as how you would have done it in dev server.
+1. Login manually to TEAMMATES on the browser used for testing. This is required as Google does not allow login by automated software.
```
[the remark this answers is not in the document]
Comment (jtankw3) [reply docs/development.md:248]: This part is for testing against production server, I believe we only use the private argument for the dev server. 
This "Login manually" part is for initially adding the cookie with the login data to the browser profile to bypass the login.

File: docs/e2e-testing.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  
+## What is E2E Testing?
+  
+E2E (End-to-end) testing is a testing methodology where the objective is to test the application as a whole.   
+- It aims to ensure all integrated components of the application work together as expected when it is being used by the end user.   
+- This is done by simulating user scenarios on the fully built product.  
+  
+E2E tests in TEAMMATES can be found in the package `teammates.e2e`. 
+
+To set up and run E2E tests, refer to this [document](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md#testing).
```
[the remark this answers is not in the document]
Comment (jtankw3) [reply docs/e2e-testing.md]: I think you are right that grouping the E2E information together is neater and more logical. Also makes the testing part for the development.md easier to understand as I believe most people will not run E2E locally (unless it starts failing and they have to debug). 

I shifted it over so you can take a look.

File: docs/e2e-testing.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+E2E tests follow this configuration:
+
+Test suite | Command | Results can be viewed in
+---|---|---
+`E2E tests` | `./gradlew e2eTests` | `{project folder}/build/reports/e2e-test-try-{n}/index.html`, where `{n}` is the sequence number of the test run
+Any individual E2E test | `./gradlew e2eTestTry1 --tests TestClassName` | `{project folder}/build/reports/e2e-test-try-1/index.html`
+
+
+- `E2E tests` will be run in their entirety once and the failed tests will be re-run a few times. 
+- Before running `E2E tests`, it is important to have the both front-end and back-end dev servers running locally first if you are testing against them.
```
Comment (wkurniawan07) [inline docs/e2e-testing.md:60]: It is mentioned like this before, but this part is not 100% accurate. Only the back-end dev server is absolutely necessary, in particular if the front-end has been built. I suggest to shorten the term to just "dev server"; anyone who is advanced enough to run E2E tests should be able to figure this out on their own.

it is important to have the ~both front-end and back-end~ dev server~s~ running locally first if you are testing against ~them~ it.

File: docs/e2e-testing.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+1. Login manually to TEAMMATES on the browser used for testing to add cookie with login details to the browser profile.
+   * This profile is added to the web driver so that E2E tests will start with user already logged in. 
+   * This is required as Google does not allow login by automated software.
+
+1. For Firefox, run the full test suite or any subset of it as how you would have done it in dev server. 
+   * Do note that the GAE daily quota is usually not enough to run the full test suite, in particular for accounts with no billing enabled.
+   
+1. For Chrome, you may have to run tests one at a time as multiple ChromeDriver instances cannot be opened with the same user data.
+
+<sup>1</sup> This setup is necessary because our test suite uses the Gmail API to access Gmail accounts used for testing (these accounts are specified in `test.properties`) to confirm that those accounts receive the expected emails from TEAMMATES.
```
Comment (wkurniawan07) [inline docs/e2e-testing.md:89]: Again, since now only one singular account is used:

to access the Gmail account~s~ used for testing (the~se~ account~s~ ~are~ is specified in `test.properties`) to confirm that th~os~e account~s~ receives the expected emails from TEAMMATES

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

**Review by wkurniawan07 — APPROVED** [review]
